### PR TITLE
Auto SSE Vectorizing

### PR DESCRIPTION
**Pull request by daschuer**
This helps on 32 bit windows version ...

### CODE REVIEW

**Comment by ywwg** [comment]
what is your evidence that it helps?

**Comment by kain88-de** [comment]
Sure this isn't making things worse on 64bit architectures?

Thanks for working on this.

Do you mind adding information about SSE and alingment in the wiki as a future reference for you and others.

**Comment by daschuer** [comment]
> What is your evidence that it helps?

I have analyzed the assembler code. It contains the vectorized loop and and index looped which is uses until the alignment suits for the 128 bit SSE registers. 
This means: aligned buffer = no slow index loop. 
If we can guarantee the the data is allays aligned we can tell GCC to remove the alignment check and the index loop. Not sure if this is worth the risk.  
However I have no test data, since the buffers on my test systems are allays aligned. It seams to be a MSVC 32 bit only issue.

> Sure this isn't making things worse on 64bit architectures?

Yes, useAlignedAlloc() will return false.

**Comment by ywwg** [comment]
Thanks for continuing to work on this sort of nitty-gritty performance stuff.  I'd really like to see profiling data on a 32 bit system confirming that this is needed and makes a significant difference.

File: src/sampleutil.cpp [hunk 1]
```diff
@@ -17,14 +17,62 @@ typedef qint32 int32_t;
...
+
+static inline bool useAlignedAlloc() {
+    // Will return true on MSVC 32 bit builds and falls for
+    // Linux 32 and 64 bit builds
+    return (sizeof(long double) == 8 && sizeof(CSAMPLE*) <= 8 &&
+            sizeof(CSAMPLE*) == sizeof(size_t));
+}
+
 // static
 CSAMPLE* SampleUtil::alloc(unsigned int size) {
```
Comment (rryan) [inline src/sampleutil.cpp]: At some point we should consider enforcing the use of SampleUtil::free (instead of delete) with a wrapper type or something. I think a raw delete on a SampleUtil::alloc'd pointer will throw an exception with the alignment tricks. 

For example: https://code.google.com/p/vertex-vertex/source/browse/trunk/util/undeletable_ptr.hpp

File: src/sampleutil.cpp [hunk 2]
```diff
@@ -17,14 +17,62 @@ typedef qint32 int32_t;
 // When changing, be carefull to not prevent the vectorizing
 // https://gcc.gnu.org/projects/tree-ssa/vectorization.html
 
+
+static inline bool useAlignedAlloc() {
+    // Will return true on MSVC 32 bit builds and falls for
```
Comment (rryan) [inline src/sampleutil.cpp]: What about mingw/clang builds on Windows?

File: src/sampleutil.cpp [hunk 3]
```diff
@@ -17,14 +17,62 @@ typedef qint32 int32_t;
...
+    // In case of sizeof(long double) = 8
+    // this code over allocates the requested buffer to be able to
+    // shift the returned pointer to a 16 byte alignment
+    // In the memory before, a pointer to the original
+    // malloced area is stored used to free the memory
+    // This code can be replaced by C11 <stdlib.h> aligned_alloc()
+    // or MSVC ::_aligned_malloc(size, alignment) and  ::_aligned_free(ptr);
+
+    if (useAlignedAlloc()) {
+        // We need to shift the alignment to 16
```
Comment (rryan) [inline src/sampleutil.cpp]: Why not just use ```posix_memalign``` / ```_aligned_malloc``` in a platform-specific ifdef for all allocations? I don't think we need to re-invent the wheel here :).

File: src/sampleutil.h [hunk 4]
```diff
@@ -58,10 +68,36 @@ class SampleUtil {
...
+        // SampleUtil::copy 661 ns 
+        //  
+        // memcpy() calls __memcpy_sse2() on 64 bit build only
+        // (not available on Debian 32 bit builds)
+        // However the Debian 32 bit memcpy() uses a SSE version of 
+        // memcpy() when called directly from Mixxx source but this 
+        // requires some checks that can be omitted when inlining the 
+        // following vectorized loop. Btw.: memcpy() calls from the Qt 
+        // library are not using SSE istructions.     
+        #ifdef __SSE__
```
Comment (rryan) [inline src/sampleutil.h]: Please don't indent ifdef's (I think it makes them stick out better).

File: src/sampleutil.cpp (hunk 1 above)
Comment (daschuer) [inline src/sampleutil.cpp]: Yes good idea. We should switch to a Sample buffer class. With this, we can issue a bunch of problems. We can use optimized Sample functions that need not check if the start and end of the buffer is aligned. We can also fix the 160000 byte overallocation this helps to keep the buffers in the CPU cache. 
But the changes will effect a lot of files, so I like to do it in a separate PR after this one is merged.

File: src/sampleutil.cpp (hunk 2 above)
Comment (daschuer) [inline src/sampleutil.cpp]: Do not know the result yet. The nice think is that useAlignedAlloc is platform and compiler independent so it should work.

File: src/sampleutil.cpp (hunk 3 above)
Comment (daschuer) [inline src/sampleutil.cpp]: This solution works the same way like the Boost generic implementation, but has a fixed 16 byte alignment. 
Calling posix_memalign is not required, because the supporting platforms should already have a 16 byte alignment by default.  
As already written in the comments we can use _aligned_malloc and _aligned_free, but this will probably use a more generic (more expensive) version of what I did and is not platform / compiler independent. 
Since I have no Win build environment to check this, I prefer to wait for C++11 and switch to aligned_alloc(), which does not require a special free.  Maybe a windows guy can jump in here.

**Comment by daschuer** [comment]
Ready for review. 
Nearly all sampleutil functions are now benefit from SSE :-)

File: src/sampleutil.cpp (hunk 3 above)
Comment (rryan) [inline src/sampleutil.cpp]: I think it's much higher risk to have this in our codebase. We should rely on the portable ways of doing this and not make assumptions about platforms. This will make it easier to port the codebase to other platforms in the future. If we want a 16-byte alignment, the POSIX-compliant way to do it is with ```posix_memalign```. On Windows platforms, it's ```_aligned_malloc```.

Even if it is generic and more expensive, why does that matter? This is SampleUtil::alloc.

File: src/sampleutil.cpp [hunk 5]
```diff
@@ -17,19 +17,69 @@ typedef qint32 int32_t;
...
+        std::free(*((void**)((void*)pBuffer) - 1));
+    } else {
+        // We are either correct aligned or on an exotic architecture
+        delete[] pBuffer;
+    }
 }
 
 // static
 void SampleUtil::applyGain(CSAMPLE* pBuffer, CSAMPLE_GAIN gain,
-        unsigned int iNumSamples) {
```
Comment (rryan) [inline src/sampleutil.cpp]: Out of curiosity, why did you change the unsigned integers to integers? Did that have an affect on the loop vectorization?

File: src/sampleutil.cpp [hunk 6]
```diff
@@ -56,76 +106,79 @@ void SampleUtil::applyRampingGain(CSAMPLE* pBuffer, CSAMPLE_GAIN old_gain,
...
+void SampleUtil::addWithGain(CSAMPLE* _RESTRICT pDest, const CSAMPLE* _RESTRICT pSrc,
+        CSAMPLE_GAIN gain, int iNumSamples) {
     if (gain == CSAMPLE_GAIN_ZERO) {
         return;
     }
 
-    // note: LOOP VECTORIZED.
-    for (unsigned int i = 0; i < iNumSamples; ++i) {
+    // note: LOOP VECTORIZED. 
+    for (int i = 0; i < (int)iNumSamples; ++i) {
```
Comment (rryan) [inline src/sampleutil.cpp]: cast should be unnecessary, right?

File: src/sampleutil.cpp [hunk 7]
```diff
@@ -213,79 +267,64 @@ void SampleUtil::copyWithRampingGain(CSAMPLE* pDest, const CSAMPLE* pSrc,
...
     CSAMPLE fAbsR = CSAMPLE_ZERO;
-    bool clipped = false;
+    CSAMPLE clipped = 0;
 
-    for (unsigned int i = 0; i < iNumSamples; i += 2) {
-        CSAMPLE absl = fabs(pBuffer[i]);
-        if (absl > CSAMPLE_PEAK) {
-            clipped = true;
-        }
+    // note: LOOP VECTORIZED.
```
Comment (rryan) [inline src/sampleutil.cpp]: maybe drop a comment about how a boolean flag prevents vectorization?

File: src/sampleutil.cpp [hunk 8]
```diff
@@ -294,72 +333,71 @@ void SampleUtil::deinterleaveBuffer(CSAMPLE* pDest1, CSAMPLE* pDest2,
...
+        int iNumSamples) {
     const CSAMPLE_GAIN cross_inc = CSAMPLE_GAIN_ONE
             / CSAMPLE_GAIN(iNumSamples / 2);
-    CSAMPLE_GAIN cross_mix = CSAMPLE_GAIN_ZERO;
-    for (unsigned int i = 0; i + 1 < iNumSamples; i += 2) {
-        pDest[i] = pSrcFadeIn[i] * cross_mix
-                + pSrcFadeOut[i] * (CSAMPLE_GAIN_ONE - cross_mix);
-        pDest[i + 1] = pSrcFadeIn[i + 1] * cross_mix
-                + pSrcFadeOut[i + 1] * (CSAMPLE_GAIN_ONE - cross_mix);
-        cross_mix += cross_inc;
```
Comment (rryan) [inline src/sampleutil.cpp]: Was this loop not vectorized? I wonder why. 

Noticing variables that are linear functions of the loop length is Compilers 101 -- a toy x86_64 compiler I wrote at university could do this :). Perhaps it's because it's a float and ```cross_inc * i``` isn't exactly equal to ```cross_inc + cross_inc + cross_inc + cross_inc + ...```

**Comment by rryan** [comment]
> If we can guarantee the the data is allays aligned we can tell GCC to remove the alignment check and the index loop. Not sure if this is worth the risk.

If we tag SampleUtil allocated buffers with a wrapper type or something then I'd be comfortable with this (along with a unit test that alloc returns a 16-byte-aligned buffer so that we may catch issues early by running the tests on new platforms). Otherwise I'd worry that someone would accidentally pass a self-allocated CSAMPLE buffer to SampleUtil.

File: src/sampleutil.cpp (hunk 3 above)
Comment (daschuer) [inline src/sampleutil.cpp]: There is no need to argue: 
We just need a windows developer who switches this to _aligned_malloc. 
posix_memalign will not help here, because it is already 16 byte aligned on platforms supporting posix_memalign

File: src/sampleutil.cpp (hunk 5 above)
Comment (daschuer) [inline src/sampleutil.cpp]: Yes. MSVC does not support vectorizing with unsigned.

File: src/sampleutil.cpp (hunk 8 above)
Comment (daschuer) [inline src/sampleutil.cpp]: Tho old version was not vectorized, the new one is. 
Induction is only recognized as a function of i

Comment (rryan) [inline src/sampleutil.cpp]: Hm! ok

File: src/sampleutil.cpp (hunk 5 above)
Comment (rryan) [inline src/sampleutil.cpp]: Interesting -- do you have a reference for that? I can't seem to find mentions of it on the web. 

Also, how are you checking what MSVC vectorizes w/o a windows build environment?

Comment (daschuer) [inline src/sampleutil.cpp]: I have not checked it. But others did:  
http://blogs.msdn.com/b/nativeconcurrency/archive/2012/05/08/auto-vectorizer-in-visual-studio-11-rules-for-loop-body.aspx

File: src/sampleutil.cpp (hunk 8 above)
Comment (daschuer) [inline src/sampleutil.cpp]: The loop counter must be ++i only, do be vectorized. i += 2 prevents vectorizing.

File: src/sampleutil.cpp (hunk 3 above)
Comment (rryan) [inline src/sampleutil.cpp]: > We just need a windows developer who switches this to _aligned_malloc. 

I think it's not that hard to change yourself:
```buffer = _aligned_malloc(size, 16);```
The build server will tell us if there's a mistake.

> posix_memalign will not help here, because it is already 16 byte aligned on platforms supporting posix_memalign

I don't think that POSIX defines that pointers must be 16-byte aligned. Does it?

If it does not then we should use posix_memalign to make sure that our goal is accomplished on other platforms. Debian builds Mixxx on a crazy number of platforms.

File: src/sampleutil.cpp (hunk 5 above)
Comment (rryan) [inline src/sampleutil.cpp]: Oh, but unsigned int matches the platform natural integer width so that article claims it will vectorize. 

Either way, what about using size_t since that more accurately represents what the argument is?

Comment (daschuer) [inline src/sampleutil.cpp]: There is no issue with int here. Qt uses is everywhere and all gcc vectorized examples are using it so it will be fine for us as well.

Comment (rryan) [inline src/sampleutil.cpp]: Ok -- there seems to be no issue with unsigned int either though which is why I was asking.

File: src/sampleutil.cpp (hunk 3 above)
Comment (daschuer) [inline src/sampleutil.cpp]: It seams to be hard to find out which platform provides what 
http://lists.boost.org/boost-commit/2013/07/46907.php

So I will only check for _MSC_VER for now.

**Comment by daschuer** [comment]
Thank you for review. 
Notes Addressed.

File: src/sampleutil.cpp (hunk 3 above)
Comment (rryan) [inline src/sampleutil.cpp]: Can we clone that Boost header into our lib folder instead? I'm trying to avoid having to maintain this code.

Comment (daschuer) [inline src/sampleutil.cpp]: Unfortunately it is not only one boost header. Copying a noter library is also not the best idea.
Adding boost dependency for this is also overdone.

If this builds on windows. Lets stick with the current version. Non of our current builds will use internal aligned allocator, which actually works by the way. The only build which uses _aligned_malloc is win 32.

Once we have c++11 there is a standard function for this. So there is no need to spend more time on this topic.

Comment (rryan) [inline src/sampleutil.cpp]: > Once we have c++11 there is a standard function for this. So there is no need to spend more time on this topic.

Ok, good point -- thanks.

**Comment by rryan** [comment]
thanks, lgtm

**Comment by daschuer** [comment]
Thank you for review!

**Comment by rryan** [comment]
Hm, it looks like the tests are failing on Windows with allocation failures:
http://builds.mixxx.org/jenkins/job/master/architecture=i386,platform=windows/291/console

**Comment by rryan** [comment]
> Hm, it looks like the tests are failing on Windows with allocation failures:
> http://builds.mixxx.org/jenkins/job/master/architecture=i386,platform=windows/291/console

I think this might be causing it? https://github.com/mixxxdj/mixxx/commit/b36239f07d81230626ddfe939c47019e26dac331

**Comment by rryan** [comment]
That seems to have fixed it.

File: src/sampleutil.cpp [hunk 9]
```diff
@@ -17,19 +17,78 @@ typedef qint32 int32_t;
...
+    // See SampleUtil::alloc() for details
+    if (useAlignedAlloc()) {
+        if (pBuffer == NULL) {
+            return;
+        }
+#if(_MSC_VER)
+        _aligned_free(pBuffer);
+#else
+        // Pointer to the original memory is stored before pBuffer
+        std::free(*((void**)((void*)pBuffer) - 1));
```
Comment (pwhelan) [inline src/sampleutil.cpp]: This line causes an error with clang on OSX. I cannot reproduce it now since my MBP is temporarily out of commission, but it has to do with pBuffer being a CSAMPLE *.

Comment (rryan) [inline src/sampleutil.cpp]: This PR broke the build on OS X -- I fixed it https://github.com/mixxxdj/mixxx/commit/096aac31004cf0939201b338b9b74d018d297295
which commit did you build from when you got the error?

Comment (pwhelan) [inline src/sampleutil.cpp]: I believe it was [e167640c9d6598254569555a5bdd44853a65d474]. I was also getting the error about 'std:free' and 'std:malloc' which adding cstdlib should fix.

Comment (pwhelan) [inline src/sampleutil.cpp]: It's likely the problem was just that, as I tried fixing 'std:free' problem by simply removing the namespace prefix (which would have resulted in an infinite loop...)
